distributed/worker.py

```diff
@@ -2910,7 +2910,14 @@ async def execute(self, key):
             try:
                 e = self.executors[executor]
                 ts.start_time = time()
-                if "ThreadPoolExecutor" in str(type(e)):
+                if iscoroutinefunction(function):
```

Comment:
The addition of this check revealed that since our `iscoroutinefunction` utility function is wrapped by `functool.lru_cache` for performance reasons, `iscoroutinefunction` raised an error when it encountered an unhashable function (which occurred in one of our tests). I pushed a commit which adds some fallback logic to make sure `iscoroutinefunction` can handle unhashable inputs